flamedisx/nest/lxe_sources.py

```diff
@@ -552,23 +555,262 @@ class nestWIMPSource(nestNRSource):
+
+        if "detector" not in kwargs:
+            kwargs["detector"] = "default"
+
+        self.energy_hist = self.get_energy_hist(
+            wimp_mass=wimp_mass,
+            sigma=sigma,
+            min_E=min_E,
+            max_E=max_E,
+            n_energy_bins=n_energy_bins,
+            min_time=min_time,
+            max_time=max_time,
+            n_time_bins=n_time_bins,
+            modulation=modulation,
+        )
 
-        self.energy_hist = pkl.load(open(os.path.join(os.path.dirname(__file__), 'wimp_spectra/WIMP_spectra.pkl'), 'rb'))[wimp_mass]
         scale = fid_mass * livetime
         self.energy_hist *= scale
 
-        self.n_time_bins = len(self.energy_hist.bin_edges[0]) - 1
-        e_centers = fd_nest.WIMPEnergySpectrum.bin_centers(self.energy_hist.bin_edges[1])
+        self.n_time_bins = len(self.energy_hist.bin_centers()[0])
+        e_centers = self.energy_hist.bin_centers()[1]
         self.energies = fd.np_to_tf(e_centers)
 
-        self.array_columns = (('energy_spectrum', len(e_centers)),)
+        self.array_columns = (("energy_spectrum", len(e_centers)),)
 
         super().__init__(*args, **kwargs)
+
+    @staticmethod
+    def get_energy_hist(
+        wimp_mass=40.0,
+        sigma=1e-45,
+        min_E=1e-2,
+        max_E=80.0,
+        n_energy_bins=800,
+        min_time="2019-09-01T08:28:00",
+        max_time="2020-09-01T08:28:00",
+        n_time_bins=25,
+        modulation=True,
+    ):
+        """
+        Function to generate the energy histogram for a given WIMP mass
+
+        Parameters
+        ----------
+        wimp_mass : float
+            Mass of the WIMP in GeV/c^2
+        min_E : float
+            Minimum energy of the histogram in keV
+        max_E : float
+            Maximum energy of the histogram in keV
+        sigma : float
+            Cross-section of the WIMP in pb
+        n_bins_energy : int
+            Number of bins for the energy histogram
+        n_bins_time : int
+            Number of bins for the time histogram.
+            To total timespan to bin is currently hardcoded to 1 year
+        modulation : bool
+            Flag to enable/disable modulation
+        """
+
+        energy_bin_edges = np.linspace(min_E, max_E, n_energy_bins + 1)
+        energy_bin_width = (energy_bin_edges[1] - energy_bin_edges[0]) * nu.keV
+        energy_values = (energy_bin_edges[1:] + energy_bin_edges[:-1]) / 2
+
+        time_bin_edges = (
+            pd.date_range(min_time, max_time, periods=n_time_bins + 1).to_julian_date()
+            - 2451545.0        # Convert to J2000
+        )
+        time_bin_width = wr.j2000_to_datetime(time_bin_edges[1]) - wr.j2000_to_datetime(time_bin_edges[0])
+        time_bin_width = time_bin_width.value * nu.ns
+        times = (time_bin_edges[:-1] + time_bin_edges[1:]) / 2
+
+        scale = time_bin_width / nu.year  # Convert from [per year] to [per time_bin_width]
```

Comment:
@lorenzomag I think we need to get rid of this. I tried to explain on Slack (maybe you missed it): whilst in principle keeping this and removing  `/ self.n_time_bins` would solve things at the level of `source.mu_before_efficiencies()`, it will mess up the tensor-based calculation of the differential rate, as the thing that is queried should be the spectrum corresponding to a given event time (such that the sum of the spectrum gives the expected counts within that energy range, after scaling by the exposure). We don't want to break compatibility with the tensor-based calculation